File: output/go/tests/47_hsm_enter_both.go

package main

import "fmt"
import "os"

type HSMEnterBothFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type HSMEnterBothFrameContext struct {
    _event  HSMEnterBothFrameEvent
    _return any
    _data   map[string]any
}

type HSMEnterBothCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *HSMEnterBothFrameEvent
    parentCompartment *HSMEnterBothCompartment
}

func newHSMEnterBothCompartment(state string) *HSMEnterBothCompartment {
    return &HSMEnterBothCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *HSMEnterBothCompartment) copy() *HSMEnterBothCompartment {
    nc := &HSMEnterBothCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type HSMEnterBoth struct {
    _state_stack []*HSMEnterBothCompartment
    __compartment *HSMEnterBothCompartment
    __next_compartment *HSMEnterBothCompartment
    _context_stack []HSMEnterBothFrameContext
    log []string
}

func NewHSMEnterBoth() *HSMEnterBoth {
    s := &HSMEnterBoth{}
    s._state_stack = make([]*HSMEnterBothCompartment, 0)
    s._context_stack = make([]HSMEnterBothFrameContext, 0)
    s.__compartment = newHSMEnterBothCompartment("Start")
    s.__next_compartment = nil
    __frame_event := HSMEnterBothFrameEvent{_message: "$>", _parameters: nil}
    __ctx := HSMEnterBothFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *HSMEnterBoth) __kernel(__e *HSMEnterBothFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &HSMEnterBothFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &HSMEnterBothFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &HSMEnterBothFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *HSMEnterBoth) __router(__e *HSMEnterBothFrameEvent) {
    switch s.__compartment.state {
    case "Start":
        s._state_Start(__e)
    case "Child":
        s._state_Child(__e)
    case "Parent":
        s._state_Parent(__e)
    }
}

func (s *HSMEnterBoth) __transition(next *HSMEnterBothCompartment) {
    s.__next_compartment = next
}

func (s *HSMEnterBoth) GoToChild() {
    __e := HSMEnterBothFrameEvent{_message: "GoToChild"}
    __ctx := HSMEnterBothFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HSMEnterBoth) GoToParent() {
    __e := HSMEnterBothFrameEvent{_message: "GoToParent"}
    __ctx := HSMEnterBothFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HSMEnterBoth) GetLog() []string {
    __e := HSMEnterBothFrameEvent{_message: "GetLog"}
    __ctx := HSMEnterBothFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result []string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.([]string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HSMEnterBoth) GetState() string {
    __e := HSMEnterBothFrameEvent{_message: "GetState"}
    __ctx := HSMEnterBothFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HSMEnterBoth) _state_Start(__e *HSMEnterBothFrameEvent) {
    if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Start"
        return
    } else if __e._message == "GoToChild" {
        __compartment := newHSMEnterBothCompartment("Child")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "GoToParent" {
        __compartment := newHSMEnterBothCompartment("Parent")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *HSMEnterBoth) _state_Child(__e *HSMEnterBothFrameEvent) {
    if __e._message == "$>" {
        s.log = append(s.log, "Child:enter")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Child"
        return
    } else if __e._message == "GoToParent" {
        __compartment := newHSMEnterBothCompartment("Parent")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *HSMEnterBoth) _state_Parent(__e *HSMEnterBothFrameEvent) {
    if __e._message == "$>" {
        s.log = append(s.log, "Parent:enter")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Parent"
        return
    } else if __e._message == "GoToChild" {
        __compartment := newHSMEnterBothCompartment("Child")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func contains(slice []string, item string) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func countOccurrences(slice []string, item string) int {
	count := 0
	for _, v := range slice {
		if v == item {
			count++
		}
	}
	return count
}

func main() {
	fmt.Println("=== Test 47: HSM Enter in Both ===")
	sm := NewHSMEnterBoth()

	// TC2.3.1: Only child's enter fires when entering child
	sm.GoToChild()
	log := sm.GetLog()
	if !contains(log, "Child:enter") {
		fmt.Printf("FAIL: Expected Child:enter, got %v\n", log)
		os.Exit(1)
	}
	if contains(log, "Parent:enter") {
		fmt.Printf("FAIL: Parent:enter should NOT fire when entering child, got %v\n", log)
		os.Exit(1)
	}
	if sm.GetState() != "Child" {
		fmt.Printf("FAIL: Expected Child, got %s\n", sm.GetState())
		os.Exit(1)
	}
	fmt.Println("TC2.3.1: Only child's enter fires when entering child - PASS")

	// TC2.3.2: Parent's enter fires only when transitioning to parent
	sm.GoToParent()
	log = sm.GetLog()
	if !contains(log, "Parent:enter") {
		fmt.Printf("FAIL: Expected Parent:enter, got %v\n", log)
		os.Exit(1)
	}
	if sm.GetState() != "Parent" {
		fmt.Printf("FAIL: Expected Parent, got %s\n", sm.GetState())
		os.Exit(1)
	}
	fmt.Println("TC2.3.2: Parent's enter fires when transitioning to parent - PASS")

	// TC2.3.3: No implicit cascading - counts should be exactly 1 each
	childCount := countOccurrences(log, "Child:enter")
	parentCount := countOccurrences(log, "Parent:enter")
	if childCount != 1 {
		fmt.Printf("FAIL: Expected exactly 1 Child:enter, got %v\n", log)
		os.Exit(1)
	}
	if parentCount != 1 {
		fmt.Printf("FAIL: Expected exactly 1 Parent:enter, got %v\n", log)
		os.Exit(1)
	}
	fmt.Println("TC2.3.3: No implicit cascading of enter handlers - PASS")

	// TC2.3.4: Going back to child fires child enter again
	sm.GoToChild()
	log = sm.GetLog()
	childCount2 := countOccurrences(log, "Child:enter")
	parentCount2 := countOccurrences(log, "Parent:enter")
	if childCount2 != 2 {
		fmt.Printf("FAIL: Expected 2 Child:enter, got %v\n", log)
		os.Exit(1)
	}
	if parentCount2 != 1 {
		fmt.Printf("FAIL: Expected still 1 Parent:enter, got %v\n", log)
		os.Exit(1)
	}
	fmt.Println("TC2.3.4: Re-entering child fires child enter again - PASS")

	fmt.Println("PASS: HSM enter in both works correctly")
}

File: output/javascript/tests/47_hsm_enter_both.js

// Test 47: HSM Enter in Both Child and Parent
//
// Tests enter handler in both levels:
// - Only child's enter fires when entering child
// - Parent's enter fires only when transitioning to parent
// - No implicit cascading of enter handlers

export class HSMEnterBothFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class HSMEnterBothFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class HSMEnterBothCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new HSMEnterBothCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class HSMEnterBoth {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;
    log = [];

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new HSMEnterBothCompartment("Start");
        this.__next_compartment = null;
        const __frame_event = new HSMEnterBothFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new HSMEnterBothFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new HSMEnterBothFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new HSMEnterBothFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    go_to_child() {
        const __e = new HSMEnterBothFrameEvent("go_to_child", null);
        const __ctx = new HSMEnterBothFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    go_to_parent() {
        const __e = new HSMEnterBothFrameEvent("go_to_parent", null);
        const __ctx = new HSMEnterBothFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    get_log() {
        const __e = new HSMEnterBothFrameEvent("get_log", null);
        const __ctx = new HSMEnterBothFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_state() {
        const __e = new HSMEnterBothFrameEvent("get_state", null);
        const __ctx = new HSMEnterBothFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_Start(__e) {
        if (__e._message === "get_log") {
            this._context_stack[this._context_stack.length - 1]._return = this.log;
            return;
        } else if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Start";
            return;
        } else if (__e._message === "go_to_child") {
            const __compartment = new HSMEnterBothCompartment("Child", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        } else if (__e._message === "go_to_parent") {
            const __compartment = new HSMEnterBothCompartment("Parent", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    _state_Parent(__e) {
        if (__e._message === "$>") {
            this.log.push("Parent:enter")
        } else if (__e._message === "get_log") {
            this._context_stack[this._context_stack.length - 1]._return = this.log;
            return;
        } else if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Parent";
            return;
        } else if (__e._message === "go_to_child") {
            const __compartment = new HSMEnterBothCompartment("Child", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    _state_Child(__e) {
        if (__e._message === "$>") {
            this.log.push("Child:enter")
        } else if (__e._message === "get_log") {
            this._context_stack[this._context_stack.length - 1]._return = this.log;
            return;
        } else if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Child";
            return;
        } else if (__e._message === "go_to_parent") {
            const __compartment = new HSMEnterBothCompartment("Parent", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }
}

function main() {
    console.log("=== Test 47: HSM Enter in Both ===");
    const s = new HSMEnterBoth();

    // TC2.3.1: Only child's enter fires when entering child
    s.go_to_child();
    let log = s.get_log();
    if (!log.includes("Child:enter")) throw new Error(`Expected Child:enter, got ${JSON.stringify(log)}`);
    if (log.includes("Parent:enter")) throw new Error(`Parent:enter should NOT fire when entering child, got ${JSON.stringify(log)}`);
    if (s.get_state() !== "Child") throw new Error(`Expected Child, got ${s.get_state()}`);
    console.log("TC2.3.1: Only child's enter fires when entering child - PASS");

    // TC2.3.2: Parent's enter fires only when transitioning to parent
    s.go_to_parent();
    log = s.get_log();
    if (!log.includes("Parent:enter")) throw new Error(`Expected Parent:enter, got ${JSON.stringify(log)}`);
    if (s.get_state() !== "Parent") throw new Error(`Expected Parent, got ${s.get_state()}`);
    console.log("TC2.3.2: Parent's enter fires when transitioning to parent - PASS");

    // TC2.3.3: No implicit cascading - counts should be exactly 1 each
    const childCount = log.filter(x => x === "Child:enter").length;
    const parentCount = log.filter(x => x === "Parent:enter").length;
    if (childCount !== 1) throw new Error(`Expected exactly 1 Child:enter, got ${JSON.stringify(log)}`);
    if (parentCount !== 1) throw new Error(`Expected exactly 1 Parent:enter, got ${JSON.stringify(log)}`);
    console.log("TC2.3.3: No implicit cascading of enter handlers - PASS");

    // TC2.3.4: Going back to child fires child enter again
    s.go_to_child();
    log = s.get_log();
    const childCount2 = log.filter(x => x === "Child:enter").length;
    const parentCount2 = log.filter(x => x === "Parent:enter").length;
    if (childCount2 !== 2) throw new Error(`Expected 2 Child:enter, got ${JSON.stringify(log)}`);
    if (parentCount2 !== 1) throw new Error(`Expected still 1 Parent:enter, got ${JSON.stringify(log)}`);
    console.log("TC2.3.4: Re-entering child fires child enter again - PASS");

    console.log("PASS: HSM enter in both works correctly");
}

main();
